### PR TITLE
rpc: calculate fees in getblock using BlockUndo data

### PR DESCRIPTION
**Pull request by robot-visions**
This change is progress towards #18771 .  It adapts the fee calculation part of #16083 and addresses some feedback.  The additional "verbosity level 3" features are planned for a future PR.

**Original PR description:**

> Using block undo data (like in #14802) we can now show fee information for each transaction in a block without the need for additional -txindex and/or a ton of costly lookups. For a start we'll add transaction fee information to getblock verbosity level 2. This comes at a negligible speed penalty (<1%).

### CODE REVIEW

**Comment by robot-visions** [comment]
The documentation for `getblock` refers to the tx format of `getrawtransaction`.  Any thoughts on what would be a reasonable approach here?

- Also update `getrawtransaction` (and the documentation) to include a fee if possible (this might be slow since the cost of loading undo data is no longer amortized over many transactions in a block)
- Update the `getblock` documentation to mention that unlike `getrawtransaction`, it might also include fees
- Leave the documentation alone and let the unfortunate user figure out the semantics themselves :)

**Comment by MarcoFalke** [comment]
Please add the right co-author to your commit, if your work is based on someone else's work

**Comment by robot-visions** [comment]
Updated, thanks!  I didn't know about `Co-authored-by` feature before.

**Comment by brakmic** [comment]
ACK 732554a8a64ad5c9d49505272d225626566898de

Built, run and tested on macOS Catalina 10.15.4

File: src/core_write.cpp [hunk 1]
```diff
@@ -186,6 +187,10 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
     entry.pushKV("locktime", (int64_t)tx.nLockTime);
 
     UniValue vin(UniValue::VARR);
+
+    CAmount nTotalIn = 0;
+    CAmount nTotalOut = 0;
```
Comment (luke-jr) [inline src/core_write.cpp:192]: Why did you revert these back to the old variable name style? (I had changed them to `amt_total_in`/`amt_total_out` in my rebase...)

File: src/core_write.cpp [hunk 2]
```diff
@@ -205,6 +210,13 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
                 }
                 in.pushKV("txinwitness", txinwitness);
             }
+            if (txundo) {
+                CAmount nPrevOut = txundo->vprevout[i].out.nValue;
+                // Sanity check to avoid overflow
+                if (MoneyRange(nPrevOut)) {
+                    nTotalIn += nPrevOut;
+                }
```
Comment (luke-jr) [inline src/core_write.cpp:218]: This seems unnecessary.

File: src/core_write.cpp [hunk 3]
```diff
@@ -224,9 +236,21 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
         ScriptPubKeyToUniv(txout.scriptPubKey, o, true);
         out.pushKV("scriptPubKey", o);
         vout.push_back(out);
+
+        // Sanity check to avoid overflow
+        if (MoneyRange(txout.nValue)) {
+            nTotalOut += txout.nValue;
+        }
```
Comment (luke-jr) [inline src/core_write.cpp:243]: This won't work right... `nTotalOut` would just be missing part of its amount, and the result calculated later will still be wrong.

File: src/core_write.cpp [hunk 4]
```diff
@@ -224,9 +236,21 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
...
     }
     entry.pushKV("vout", vout);
 
+    // For coinbase transactions we have txundo == nullptr
+    if (txundo) {
+        CAmount nFee = nTotalIn - nTotalOut;
+        assert(MoneyRange(nFee));
```
Comment (luke-jr) [inline src/core_write.cpp:250]: `CHECK_NONFATAL` in RPC code

File: src/core_write.cpp [hunk 5]
```diff
@@ -224,9 +236,21 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
...
+        // Sanity check to avoid overflow
+        if (MoneyRange(txout.nValue)) {
+            nTotalOut += txout.nValue;
+        }
     }
     entry.pushKV("vout", vout);
 
+    // For coinbase transactions we have txundo == nullptr
+    if (txundo) {
+        CAmount nFee = nTotalIn - nTotalOut;
```
Comment (luke-jr) [inline src/core_write.cpp:249]: `fees`

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -143,16 +143,20 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
     result.pushKV("versionHex", strprintf("%08x", block.nVersion));
     result.pushKV("merkleroot", block.hashMerkleRoot.GetHex());
     UniValue txs(UniValue::VARR);
-    for(const auto& tx : block.vtx)
-    {
-        if(txDetails)
-        {
+    if (txDetails) {
+        CBlockUndo blockUndo;
+        bool hasUndo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp:148]: `have_undo`

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -143,16 +143,20 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
...
+        for (size_t i = 0; i < block.vtx.size(); ++i) {
+            const auto& tx = block.vtx[i];
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp:150]: `.at(i)`

**Review by luke-jr — COMMENTED** [review]
Looks like it was based on the original PR rather than my rebase, and therefore missed some fixups.

**Comment by luke-jr** [comment]
Proposed update: https://github.com/bitcoin/bitcoin/compare/master...luke-jr:getblock_fees-0.20

File: src/core_write.cpp (hunk 2 above)
Comment (robot-visions) [reply src/core_write.cpp:218]: Yeah, I agree it seems unnecessary.  I added this in order to fix a failing fuzz test.

File: src/core_write.cpp (hunk 1 above)
Comment (robot-visions) [reply src/core_write.cpp:192]: No particular reason; I'd be happy with either way.

**Comment by robot-visions** [comment]
Thanks for the review and the update @luke-jr !  I'm happy to go with your proposed update (we can close this PR).

Quick notes:
- I didn't include the verbosity 3 changes because I wanted to move things over incrementally
- In your updates it might still be necessary to introduce some way to prevent the fuzz tests from failing (it was causing an integer overflow when adding a large value to `amt_total_out`); I added a halfhearted check (which you correctly pointed out doesn't always work) but maybe it makes more sense to just update the fuzz test

**Comment by luke-jr** [comment]
We don't need to close this PR, you're doing fine.

If you want to adopt my branch as-is, you can just checkout yours, `git reset --hard 0b46eb7ac63`, and re-push it here.

File: test/functional/rpc_getblock_fee.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        self.num_nodes = 1
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        node = self.nodes[0]
+        address = node.getnewaddress()
+        node.generatetoaddress(101, address)
+        txid = node.sendtoaddress(address, 1)
```
Comment (luke-jr) [inline test/functional/rpc_getblock_fee.py:25]: `txid` is unused.

File: src/core_write.cpp [hunk 9]
```diff
@@ -224,9 +234,18 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
         ScriptPubKeyToUniv(txout.scriptPubKey, o, true);
         out.pushKV("scriptPubKey", o);
         vout.push_back(out);
+
+        amt_total_out += txout.nValue;
```
Comment (luke-jr) [inline src/core_write.cpp:238]: The fuzzer can trigger undefined behaviour here with an overflow, which causes CI to fail even though it's harmless. Not sure what the solution should be - seems like overkill to actually prevent.

Comment (practicalswift) [reply src/core_write.cpp:238]: @luke-jr Functions are assumed to be robust as long as they are fed with inputs fulfilling their documented preconditions, so I think the only robust alternative to preventing it is to make the implicit precondition explicit by documenting it. However, in this case it seems easy to just use `MoneyRange(…)` and be done with it, no? :)

File: src/core_write.cpp [hunk 10]
```diff
@@ -209,6 +215,11 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
                 txinwitness.push_back(HexStr(item));
             }
             in.pushKV("txinwitness", txinwitness);
+            if (txundo) {
+                const CTxOut& txout = txundo->vprevout[i].out;
+                // Suppress fuzz test failures by ignoring nonsense values
```
Comment (jnewbery) [inline src/core_write.cpp:220]: It seems strange to me to have code in the produce that exists only to work around shortcomings in the fuzz tests. Which fuzz test fails if you don't do this? (the only fuzz tester that calls `TxToUniv()` doesn't provide a `txundo` param).

**Review by jnewbery — COMMENTED** [review]
Looks good. It'd be nice if the test checked that the fee returned was correct rather than just testing that any value is returned.

File: src/core_write.cpp (hunk 10 above)
Comment (MarcoFalke) [reply src/core_write.cpp:220]: Agree with @jnewbery . It is impossible to hit this condition in production, so an assert or just nothing at all might be more appropriate.

Comment (practicalswift) [reply src/core_write.cpp:220]: I suggest stating the pre-condition in the form of an assertion or documentation. Then it is the fuzzer's responsibility the fulfil all explicitly stated pre-conditions.

File: test/functional/rpc_getblock_fee.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+'''Test getblock rpc fee calculation.
+'''
+
+import os
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import get_datadir_path
+
+
+class GetblockFeeTest(BitcoinTestFramework):
```
Comment (fjahr) [inline test/functional/rpc_getblock_fee.py:14]: I think `rpc_blockchain.py` is the natural place for these tests. A new file is not necessary for such a specific test IMO.

**Review by fjahr — COMMENTED** [review]
Concept ACK

**Comment by troygiorshev** [comment]
Approach ACK.

Reviewed changes, ran tests.  Need to manually test further. 💻 

I agree with the previous review comments.

Feel free to add a commit at the beginning cleaning up the code around your changes!
For example:
* Fixing the strange lack of braces around the body of the `if` [here in core_write.cpp](https://github.com/robot-visions/bitcoin/blob/1298f0fca79803f3bc91c052f901a4df63ae343a/src/core_write.cpp#L202)
* Changing to [brace initialization](https://stackoverflow.com/questions/18222926/why-is-list-initialization-using-curly-braces-better-than-the-alternatives) for lines like [this one](https://github.com/robot-visions/bitcoin/blob/1298f0fca79803f3bc91c052f901a4df63ae343a/src/core_write.cpp#L194)

File: test/functional/rpc_getblock_fee.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+'''Test getblock rpc fee calculation.
+'''
```
[the remark this answers is not in the document]
Comment (glozow) [reply test/functional/rpc_getblock_fee.py:6]: nit: I think we also usually do double quotes? `"""`

**Comment by Emzy** [comment]
I get following error:
```
$ bitcoin-cli getblock "$(bitcoin-cli getblockhash 650684)" 2
error code: -1
error message:
core_write.cpp:251 (TxToUniv)
Internal bug detected: 'MoneyRange(fee)'
You may report this issue here: https://github.com/bitcoin/bitcoin/issues
```
A block with only a Coinbase transaction works, for example 650026.

File: src/rpc/blockchain.cpp [hunk 13]
```diff
@@ -162,16 +162,21 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
     result.pushKV("versionHex", strprintf("%08x", block.nVersion));
     result.pushKV("merkleroot", block.hashMerkleRoot.GetHex());
     UniValue txs(UniValue::VARR);
-    for(const auto& tx : block.vtx)
-    {
-        if(txDetails)
-        {
+    if (txDetails) {
+        CBlockUndo blockUndo;
+        const bool have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
```
[the remark this answers is not in the document]
Comment (robot-dreams) [reply src/rpc/blockchain.cpp]: Great idea!

Yes, according to https://github.com/bitcoin/bitcoin/blob/1b313cacc99a1b372238f9036abed5491f9d28f7/src/validation.h#L370 it looks like block and undo data are always deleted together.

I definitely might be missing something, but from a quick search I couldn't find any other reason we'd delete `rev` but not `blk` files.

[the remark this answers is not in the document]
Comment (sipa) [reply src/rpc/blockchain.cpp]: I think it's not unreasonable to delete undo data while keeping block data (that means you could still serve it to other syncing nodes, and rescan, but can't reorg out of it). That just was never implemented.

[the remark this answers is not in the document]
Comment (robot-dreams) [reply src/rpc/blockchain.cpp]: OK, you both make some great points. I can think of three approaches at the moment:

- Use `GetUndoChecked` which means if someone deletes undo data in the future, they'll cause this RPC to fail
- Silently leave out the fee if block data is present but undo data is missing
- Add some verbose documentation that says "we'll include the fee if we have the data needed to efficiently calculate it"

Do any of these approaches seem reasonable to both of you?

**Comment by Emzy** [comment]
> I get following error:
> 
> ```
> $ bitcoin-cli getblock "$(bitcoin-cli getblockhash 650684)" 2
> error code: -1
> error message:
> core_write.cpp:251 (TxToUniv)
> Internal bug detected: 'MoneyRange(fee)'
> You may report this issue here: https://github.com/bitcoin/bitcoin/issues
> ```
> 
> A block with only a Coinbase transaction works, for example 650026.

8a7229e2e5c306e86a6a589c1ab0e4c58ea49545 fixed it.

File: src/rpc/blockchain.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (fjahr) [reply src/rpc/blockchain.cpp]: As @sipa says it's not implemented and I think there is no one working on something like that (Side-note: I will look into that now :) ). So I would say the first option would still be my first choice if we could have a test that would reliably fail when the change comes. But that seems unrealistic, at least if it is implemented the way I envision it now and generally it would be poking around in the dark. So, I guess I would be happy if you add some docs but you don't have to since with the current state of the code the fee should be there reliably.

File: src/rpc/blockchain.cpp [hunk 14]
```diff
@@ -926,6 +931,7 @@ static RPCHelpMan getblock()
                         {RPCResult::Type::OBJ, "", "",
                         {
                             {RPCResult::Type::ELISION, "", "The transactions in the format of the getrawtransaction RPC. Different from verbosity = 1 \"tx\" result"},
+                            {RPCResult::Type::NUM, "fee", "The transaction fee in " + CURRENCY_UNIT + ", omitted if efficient calculation isn't possible (due to pruning)"},
```
Comment (fjahr) [inline src/rpc/blockchain.cpp:934]: Hm, I think since with the currently available version of pruning the block wouldn't be found anyway this would be slightly better and also more precise:
```suggestion
                            {RPCResult::Type::NUM, "fee", "The transaction fee in " + CURRENCY_UNIT + ", omitted if efficient calculation isn't possible (due to block undo data being unavailable)"},
```

File: src/core_write.cpp [hunk 15]
```diff
@@ -189,13 +191,20 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
     entry.pushKV("weight", GetTransactionWeight(tx));
     entry.pushKV("locktime", (int64_t)tx.nLockTime);
 
-    UniValue vin(UniValue::VARR);
+    UniValue vin{UniValue::VARR};
+
+    // If available, use Undo data to calculate the fee. Note that txundo == nullptr
+    // for coinbase transactions and for transactions in pruned blocks.
```
Comment (fjahr) [inline src/core_write.cpp:197]: I think it's slightly better to be a bit more precise here. There may be other reason (in the future) why there is no undo data.
```suggestion
    // for coinbase transactions and for transactions where undo data is unavailable.
```

**Review by fjahr — COMMENTED** [review]
utACK 8ccfbde6fafa9f862bc21a8337a86f0cc465967f

Feel free to ignore my comments unless you have to make other changes.

File: src/rpc/blockchain.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/blockchain.cpp:167]: I think either leaving it like it is now, or returning an error when the undo file is not available but the block file is (since that's currently an unexpected condition) are both fine.

File: src/rpc/blockchain.cpp [hunk 16]
```diff
@@ -926,6 +931,7 @@ static RPCHelpMan getblock()
                         {RPCResult::Type::OBJ, "", "",
                         {
                             {RPCResult::Type::ELISION, "", "The transactions in the format of the getrawtransaction RPC. Different from verbosity = 1 \"tx\" result"},
+                            {RPCResult::Type::NUM, "fee", "The transaction fee in " + CURRENCY_UNIT + ", omitted if efficient calculation isn't possible (due to unavailable block undo data)"},
```
Comment (jnewbery) [inline src/rpc/blockchain.cpp:934]: I think this would be fine as just "omitted if block undo data is not available". I don't think users would care about or understand what you mean by 'efficient calculation'.

File: test/functional/rpc_blockchain.py [hunk 17]
```diff
@@ -352,6 +355,42 @@ def assert_waitforheight(height, timeout=2):
         assert_waitforheight(current_height)
         assert_waitforheight(current_height + 1)
 
+    def _test_getblock(self):
+        if not self.is_wallet_compiled():
```
Comment (jnewbery) [inline test/functional/rpc_blockchain.py:359]: Did you look at using miniwallet.py to manually create the transactions? That would avoid a dependency on a compiled wallet to test this functionality.

File: test/functional/rpc_blockchain.py [hunk 18]
```diff
@@ -352,6 +355,42 @@ def assert_waitforheight(height, timeout=2):
...
+        blockhash = node.generatetoaddress(1, address)[0]
+
+        self.log.info("Test that getblock with verbosity 1 doesn't include fee")
+        block = node.getblock(blockhash, 1)
+        assert 'fee' not in block['tx'][1]
+
+        self.log.info('Test that getblock with verbosity 2 includes expected fee')
+        block = node.getblock(blockhash, 2)
+        assert 'fee' in block['tx'][1]
+        assert_equal(block['tx'][1]['fee'], Decimal('0.00004380'))
```
Comment (jnewbery) [inline test/functional/rpc_blockchain.py:375]: Unless you're setting amounts/fee explicitly, this feels brittle (i.e. this test would break if an unrelated change was made in how the wallet code sets fees).

File: test/functional/rpc_blockchain.py (hunk 17 above)
Comment (robot-dreams) [reply test/functional/rpc_blockchain.py:359]: Thanks, I didn't know about `MiniWallet` before. That is very useful!

**Comment by jnewbery** [comment]
utACK dd4ec0704c66c23d207bb7394823469a1b59b060

Since the first commit is very different from the version in 16083, you should feel free to set yourself as the commit author and add felix as a co-author (or just add yourself as co-author).

Also feel free to set the review comments above as 'resolved' if you think they're addressed. I went through and resolved my own review comments, but didn't resolve comments from other reviewers. Setting all comments to resolved signals to other reviewers that you've addressed all the outstanding issues.

**Comment by luke-jr** [comment]
tACK 66d012ad7f9381bacfc9b8388fa2ebf82cb86c9e

**Comment by fjahr** [comment]
tACK 66d012ad7f9381bacfc9b8388fa2ebf82cb86c9e

Changes since last review were a small refactoring of the functional test using MiniWallet and some improvements on comments/docs.

File: test/functional/rpc_blockchain.py [hunk 19]
```diff
@@ -352,6 +356,46 @@ def assert_waitforheight(height, timeout=2):
...
+        blockhash = node.generate(1)[0]
+
+        self.log.info("Test that getblock with verbosity 1 doesn't include fee")
+        block = node.getblock(blockhash, 1)
+        assert 'fee' not in block['tx'][1]
+
+        self.log.info('Test that getblock with verbosity 2 includes expected fee')
+        block = node.getblock(blockhash, 2)
+        tx = block['tx'][1]
+        assert 'fee' in tx
```
Comment (MarcoFalke) [inline test/functional/rpc_blockchain.py:379]: no need for this check. The next line would already fail with `KeyError`

File: test/functional/rpc_blockchain.py [hunk 20]
```diff
@@ -352,6 +356,46 @@ def assert_waitforheight(height, timeout=2):
...
+        assert_equal(tx['fee'], tx['vsize'] * fee_per_byte)
+
+        self.log.info("Test that getblock with verbosity 2 still works with pruned Undo data")
+        datadir = get_datadir_path(self.options.tmpdir, 0)
```
Comment (MarcoFalke) [inline test/functional/rpc_blockchain.py:383]: no need for the import `node.datadir` should suffice

**Review by maflcko — APPROVED** [review]
Untitled pnggg


review ACK 66d012ad7f9381bacfc9b8388fa2ebf82cb86c9e 🗜

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 66d012ad7f9381bacfc9b8388fa2ebf82cb86c9e 🗜
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiUbgwAzPzceMugFaUVExZY36XvjsXenEtSk7a/xjrn8+SkZOd1gPtb852CYkQF
CuWb42KnECcFnSw+BvesBuHqn6lcOdHlD8yqNaA3IJdW1UJREN0ik8rScuBnNigS
7PrsiSJVESlXP16YcvlcFl/z/OmZ15tY+p+6IIed5/tp1CC5SYvkx2qUgLFXomU3
pbiuzE6rePULNjthMcS+CO8/8id2hPT4UoJdRxEZJ5pAqaIAtdwe0XFAUEFt6jwM
2gkafeYnfk3c/lz/oIschVRrwFI/Hff7H/yuVPDzXebzQvHyMOATSyguB4xSVtyD
tXOUlVovzdCkm6pTXn2gFLm59NbPDW0VbfJcFrhTEFdFV70MQ3scqgckTu74TZ4p
cCQtQxt+bL4yHh3/CaiFon/OCK2kusAnajjqXCjwp8fmhbve5gI1t9QZXLy2QN9i
4cs6ppl+i/e+Ts0yl0xziwecso7Xcx3UwXbzfNK+UQ1p8UhSLm8xaU0LVu6aNEcz
+E+TevCn
=deRF
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3dfd17a565ae7ddbda9eaa8abb03935bc4380ceae2f4f25005f5f9c93ad5a7c8  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083dfd17a565ae7ddbda9eaa8abb03935bc4380ceae2f4f25005f5f9c93ad5a7c8f0101f9c883657c75d59ab33fa40582b076808fff010060a05c4f6e27ea4bff407287a3ab54608f1045fe4a626f0087202ea3c0ac089fb0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d75d6a4068db96c10733ccae621d08f408f020f05f87cf2533d92954994ce8bcea56d6b9fc4d9b3b1358c2543189f0d3ed6c6508f12002681ffd3a08f71223768a725356e42de1d19c1976b8102d1824b2d2610b6caa08f0203e9f1e5dfbaa6099a2030f6a88563ebecfea8ccf1d50f36972cb49d1513ce95408f1045fe4a626f0083d6c4857d4f690930083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010eb6254e76ba212a86dd4616e93f6192208f02007f969677ce7dd8796cfb548988c65cf1111ada746c481a002159b4bfcd9a1a108f0206f3f00aca032fc597421da85d51cb4c8ba336788ad751d2c64839f996c739cca08f1045fe4a626f0082db5669873cede530083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010fd2e13139389a6be8df87d36931235c908f1045fe4a626f0081c1855e3d071622b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
